### PR TITLE
[ROCm][Quantization] add quark w4a8 mxfp4_fp8 for LinearLayer

### PR DESCRIPTION
**Pull request by divakar-amd**
weights: mxfp4 with static scales
activations: fp8

1. Supports Eager + Torch.compile mode
2. Supports Emulation mode
3. Updated test results below. ~~Tested on [this script](https://gist.github.com/divakar-amd/aad32a1fe87088392fb1f44dff26219b#file-quark_quantize_gpt-oss_wmxfp4_afp8_linearlayeronly-sh) for quantized gpt-oss LinearLayer only (with MoE quant_confit set to None)~~

~~- [x] Remove temporary hack: Force MoE to use un-quantized path instead of Quark [at this location.](https://github.com/vllm-project/vllm/pull/35316/changes#diff-ce5082bac9eea30a0145ce0a72576a8a4fbceae025fcdc98e714f976072c6704R177)~~

~~#### Testing with MoE quant_config=None~~

## Testing e2e after [30357](https://github.com/vllm-project/vllm/pull/30357) merged
Edit: Update quantised model link. Here's how the model was quantised- [LINK](https://gist.github.com/divakar-amd/483ca493a482c102811a29c34ac4a405)
**Server:**
`VLLM_ROCM_USE_AITER=1 VLLM_ROCM_USE_AITER_UNIFIED_ATTENTION=1 vllm serve /data/models/gpt-oss-120b-w-mxfp4-a-fp8-kvfp8-no_lmhead_router/`

**Client:**

> curl http://localhost:8000/v1/chat/completions   -H "Content-Type: application/json"   -d '{
    "model": "/data/models/gpt-oss-120b-w-mxfp4-a-fp8-kvfp8-no_lmhead_router/",
    "messages": [
      {"role": "system", "content": "You are a helpful assistant."},
      {"role": "user", "content": "Explain briefly the process of photosynthesis."}
    ],
    "max_tokens": 100,
    "temperature": 0.7
  }'

**Output:**

> {"id":"chatcmpl-941316cb808bf653","object":"chat.completion","created":1773175378,"model":"/data/models/gpt-oss-120b-w-mxfp4-a-fp8-kvfp8-no_lmhead_router/","choices":[{"index":0,"message":{"role":"assistant","content":"**Photosynthesis – a quick overview**\n\n1. **Where it happens:** In the chloroplasts of plant (and some algae) cells, especially in the thylakoid","refusal":null,"annotations":null,"audio":null,"function_call":null,"tool_calls":[],"reasoning":"User asks: \"Explain briefly the process of photosynthesis.\" So they want a concise explanation. Should describe light-dependent reactions and Calvin cycle, main steps, inputs and outputs. Provide a brief overview. Probably a paragraph or bullet points. Let's produce a succinct yet clear explanation."},"logprobs":null,"finish_reason":"length","stop_reason":null,"token_ids":null}],"service_tier":null,"system_fingerprint":null,"usage":{"prompt_tokens":85,"total_tokens":185,"completion_tokens":100,"prompt_tokens_details":null},"prompt_logprobs":null,"prompt_token_ids":null,"kv_transfer_params":null}

**Lm-eval**

> lm_eval --model local-completions --model_args "base_url=http://localhost:8000/v1/completions,pretrained=/data/models/gpt-oss-120b-w-mxfp4-a-fp8-kvfp8-no_lmhead_router/,tensor_parallel_size=1,add_bos_token=true,trust_remote_code=true,max_gen_toks=2048" --tasks gsm8k --num_fewshot 5 --batch_size 64 --limit 250 --apply_chat_template --seed 42

|Tasks|Version|     Filter     |n-shot|  Metric   |   |Value|   |Stderr|
|-----|------:|----------------|-----:|-----------|---|----:|---|-----:|
|gsm8k|      3|flexible-extract|     5|exact_match|↑  |0.896|±  |0.0193|
|     |       |strict-match    |     5|exact_match|↑  |0.040|±  |0.0124|

## Testing emulation on gfx942

<details>
  <summary>**Quark quantising script for llama-2-7b**</summary>

  > #/bin/bash
     exclude_layers="\*lm_head* \*router*"
      python3 quantize_quark.py \
      --model_dir /data/models/llama-2-7b-chat-hf \
     --quant_scheme mxfp4_fp8 \
     --kv_cache_dtype fp8 \
  --exclude_layers $exclude_layers \
  --num_calib_data 512 \
  --output_dir /data/models/llama-2-7b-chat-hf-a8w4 \
  --model_export hf_format \
  --multi_gpu
</details>

**Server**
> vllm serve /data/models/llama-2-7b-chat-hf-a8w4/  --enforce_eager

**Client**
> curl http://localhost:8000/v1/chat/completions -H "Content-Type: application/json" -d '{
"model": "/data/models/llama-2-7b-chat-hf-a8w4/",
"messages": [
{"role": "system", "content": "You are a helpful assistant."},
{"role": "user", "content": "Explain briefly the process of photosynthesis."}
],
"max_tokens": 100,
"temperature": 0.7
}'

**Output**
> {"id":"chatcmpl-b40772573111c4cf","object":"chat.completion","created":1773186698,"model":"/data/models/llama-2-7b-chat-hf-a8w4/","choices":[{"index":0,"message":{"role":"assistant","content":"  Of course! Photosynthesis is the process by which plants, algae, and some bacteria convert light energy from the sun into chemical energy in the form of organic compounds, such as glucose. Here's a brief overview of the process:\n\n\n1. Light-dependent reactions: In the light-dependent reactions, chlorophyll (a green pigment) absorbs light energy from the sun, which excites electrons","refusal":null,"annotations":null,"audio":null,"function_call":null,"tool_calls":[],"reasoning":null},"logprobs":null,"finish_reason":"length","stop_reason":null,"token_ids":null}],"service_tier":null,"system_fingerprint":null,"usage":{"prompt_tokens":37,"total_tokens":137,"completion_tokens":100,"prompt_tokens_details":null},"prompt_logprobs":null,"prompt_token_ids":null,"kv_transfer_params":null}root@banff-cyxtera-s73-5:/Projects/VLLM_DIR/rocm_vllm#

### CODE REVIEW

File: vllm/model_executor/models/gpt_oss.py [hunk 1]
```diff
@@ -155,7 +155,7 @@ def __init__(
         super().__init__()
 
         config = vllm_config.model_config.hf_config
-        quant_config = vllm_config.quant_config
+        # quant_config = vllm_config.quant_config # [DV] Temp. Remove this.
```
[the remark this answers is not in the document]
Comment (divakar-amd) [reply vllm/model_executor/models/gpt_oss.py]: I know... chill!

File: vllm/model_executor/models/gpt_oss.py [hunk 2]
```diff
@@ -174,7 +174,7 @@ def __init__(
             intermediate_size=config.intermediate_size,
             reduce_results=True,
             renormalize=True,
-            quant_config=quant_config,
+            quant_config=None,  # [DV] Remove this. Forces w4a8 for LinearLayer only
```
[the remark this answers is not in the document]
Comment (divakar-amd) [reply vllm/model_executor/models/gpt_oss.py]: I know... will be removed, chill

File: vllm/model_executor/layers/quantization/quark/quark.py [hunk 3]
```diff
@@ -320,6 +321,31 @@ def _is_static_tensor_w8a8(
...
+            weight_quant.get("dtype") == "fp4"
+            and weight_quant.get("qscheme") == "per_group"
+            and weight_quant.get("group_size") == 32
+            and weight_quant.get("scale_format") == "e8m0"
+            and not weight_quant.get("is_dynamic")
+        )
+
+        is_input_fp8 = (
+            input_quant.get("dtype") == "fp8_e4m3"
+            and input_quant.get("qscheme") in ["per_tensor", "per_channel"]
```
Comment (BowenBao) [inline vllm/model_executor/layers/quantization/quark/quark.py]: suggest removing 'per_channel'. although the kernel supports per-channel scales, typically in quantization we don't produce per_channel _static_ scales for input. For fine grained input quant like per-channel and per-group, we use dynamic quantization.

File: vllm/model_executor/layers/quantization/quark/schemes/quark_w4a8_mxfp4_fp8.py [hunk 4]
```diff
@@ -0,0 +1,208 @@
...
+                "Dynamic FP8 activation quantization is not yet supported "
+                "for W4A8. The current implementation expects static per-tensor "
+                "FP8 scales stored in the checkpoint."
+            )
+
+        self.use_aiter_kernel = (
+            is_aiter_found_and_supported() and self.is_static_input_scheme
+        )
+
+        if not self.use_aiter_kernel:
```
Comment (BowenBao) [inline vllm/model_executor/layers/quantization/quark/schemes/quark_w4a8_mxfp4_fp8.py:77]: does this kernel require gfx950?

File: vllm/model_executor/models/gpt_oss.py
Comment (BowenBao) [inline vllm/model_executor/models/gpt_oss.py:1]: remove debug lines

**Review by BowenBao — APPROVED** [review]
LGTM, some minor comments

File: vllm/model_executor/layers/quantization/quark/quark.py (hunk 3 above)
Comment (divakar-amd) [reply vllm/model_executor/layers/quantization/quark/quark.py]: Thanks. Updated.

File: vllm/model_executor/layers/quantization/quark/schemes/quark_w4a8_mxfp4_fp8.py (hunk 4 above)
Comment (divakar-amd) [reply vllm/model_executor/layers/quantization/quark/schemes/quark_w4a8_mxfp4_fp8.py:77]: Good catch, thanks. Fixed

File: vllm/model_executor/layers/quantization/quark/schemes/quark_w4a8_mxfp4_fp8.py [hunk 5]
```diff
@@ -0,0 +1,214 @@
...
+        self,
+        layer: torch.nn.Module,
+        x: torch.Tensor,
+        bias: torch.Tensor | None = None,
+    ) -> torch.Tensor:
+        M = x.shape[0]
+        out_dtype = x.dtype if self.out_dtype is None else self.out_dtype
+
+        input_scale = layer.input_scale
+        finfo = torch.finfo(torch.float8_e4m3fn)  # [-448, 448]
```
Comment (tjtanaa) [inline vllm/model_executor/layers/quantization/quark/schemes/quark_w4a8_mxfp4_fp8.py]: Use this function `get_fp8_min_max()`  from `vllm.model_executor.layers.quantization.utils.quant_utils`

https://github.com/vllm-project/vllm/blob/156e33553ccdba940fec83a720290b30d2686ee8/vllm/model_executor/layers/quantization/utils/quant_utils.py#L25

Do not call `get_fp8_min_max()` in forwardpass. Store the min max value during initializtion.

File: vllm/model_executor/layers/quantization/quark/schemes/quark_w4a8_mxfp4_fp8.py [hunk 6]
```diff
@@ -0,0 +1,214 @@
...
+        x_fp8 = (x / input_scale).clamp(finfo.min, finfo.max).to(torch.float8_e4m3fn)
```
Comment (tjtanaa) [inline vllm/model_executor/layers/quantization/quark/schemes/quark_w4a8_mxfp4_fp8.py]: use `current_platform.fp8_dtype()` if we expect emulation also works for MI300X.

Do not call `current_platform.fp8_dtype()` in forwardpass. Store the min max value during initializtion.

File: vllm/model_executor/layers/quantization/quark/schemes/quark_w4a8_mxfp4_fp8.py [hunk 7]
```diff
@@ -0,0 +1,214 @@
...
+
+logger = init_logger(__name__)
+
+
+__all__ = ["QuarkW4A8_MXFP4_FP8"]
+
+OCP_MX_BLOCK_SIZE = 32
+
+
+class QuarkW4A8_MXFP4_FP8(QuarkScheme):
```
Comment (tjtanaa) [inline vllm/model_executor/layers/quantization/quark/schemes/quark_w4a8_mxfp4_fp8.py:33]: Is this supported on `gfx942` ?

File: vllm/_aiter_ops.py [hunk 8]
```diff
@@ -861,6 +861,39 @@ def _rocm_aiter_triton_add_rmsnorm_pad_fake(
...
+    x: torch.Tensor,
+    w: torch.Tensor,
+    x_scales: torch.Tensor,
+    w_scales: torch.Tensor,
+    out_dtype: torch.dtype,
+) -> torch.Tensor:
+    from aiter.ops.triton.gemm_a8wfp4 import gemm_a8wfp4
+
+    M, N = x.shape[0], w.shape[0]
+    y = torch.zeros(M, N, dtype=out_dtype, device=x.device)
```
Comment (tjtanaa) [inline vllm/_aiter_ops.py]: @divakar-amd can we use `torch.empty` here for performance benefit? This is a question because I remember on ROCm, some other ops seemed cause accuracy issue occasionally if we initialize it with `torch.empty`. 
Can you find out if we can swap it with `torch.empty`? The allocation speed of `torch.zeros` is order of magnitudes higher slower than `torch.empty`.

Since the `_rocm_aiter_gemm_a8wfp4_impl` is wrapped under `direct_register_customop`, the `y` tensor is not capture in torch compile graph and cudagraph.

File: vllm/model_executor/layers/quantization/quark/schemes/quark_w4a8_mxfp4_fp8.py [hunk 9]
```diff
@@ -0,0 +1,214 @@
...
+
+        weight_dq = dequant_mxfp4(
+            layer.weight,
+            layer.weight_scale,
+            x.dtype,
+        )
+
+        input_scale = layer.input_scale
+        finfo = torch.finfo(torch.float8_e4m3fn)  # [-448, 448]
+        x_fp8 = (x / input_scale).clamp(finfo.min, finfo.max).to(torch.float8_e4m3fn)
```
Comment (tjtanaa) [inline vllm/model_executor/layers/quantization/quark/schemes/quark_w4a8_mxfp4_fp8.py]: use `current_platform.fp8_dtype()` if we expect emulation also works for MI300X.

Do not call `current_platform.fp8_dtype()` in forwardpass. Store the min max value during initializtion.

File: vllm/model_executor/layers/quantization/quark/schemes/quark_w4a8_mxfp4_fp8.py [hunk 10]
```diff
@@ -0,0 +1,214 @@
...
+        )
+
+        weight_dq = dequant_mxfp4(
+            layer.weight,
+            layer.weight_scale,
+            x.dtype,
+        )
+
+        input_scale = layer.input_scale
+        finfo = torch.finfo(torch.float8_e4m3fn)  # [-448, 448]
```
Comment (tjtanaa) [inline vllm/model_executor/layers/quantization/quark/schemes/quark_w4a8_mxfp4_fp8.py]: Use this function `get_fp8_min_max()`  from `vllm.model_executor.layers.quantization.utils.quant_utils`

https://github.com/vllm-project/vllm/blob/156e33553ccdba940fec83a720290b30d2686ee8/vllm/model_executor/layers/quantization/utils/quant_utils.py#L25

Do not call `get_fp8_min_max()` in forwardpass. Store the min max value during initializtion.

File: vllm/model_executor/layers/quantization/quark/schemes/quark_w4a8_mxfp4_fp8.py [hunk 11]
```diff
@@ -0,0 +1,214 @@
...
+    def apply_weights(
+        self,
+        layer: torch.nn.Module,
+        x: torch.Tensor,
+        bias: torch.Tensor | None = None,
+    ) -> torch.Tensor:
+        if self.use_aiter_kernel:
+            return self._apply_aiter_kernel(layer, x, bias)
+        else:
+            return self._apply_emulation(layer, x, bias)
```
Comment (tjtanaa) [inline vllm/model_executor/layers/quantization/quark/schemes/quark_w4a8_mxfp4_fp8.py:172]: Is this `_apply_emulation` a function that can also be run on `gfx942` ?

File: vllm/_aiter_ops.py (hunk 8 above)
Comment (divakar-amd) [reply vllm/_aiter_ops.py]: Thanks for mentioning this. I used` torch.zeros` for the very same concerns. However, I just swapped & tested with `torch.empty` and we are good to go. (Used lm-eval and client output to check the correctness). I'll update the code.

File: vllm/model_executor/layers/quantization/quark/schemes/quark_w4a8_mxfp4_fp8.py (hunk 5 above)
Comment (divakar-amd) [reply vllm/model_executor/layers/quantization/quark/schemes/quark_w4a8_mxfp4_fp8.py]: Done

File: vllm/model_executor/layers/quantization/quark/schemes/quark_w4a8_mxfp4_fp8.py (hunk 6 above)
Comment (divakar-amd) [reply vllm/model_executor/layers/quantization/quark/schemes/quark_w4a8_mxfp4_fp8.py]: Done

File: vllm/model_executor/layers/quantization/quark/schemes/quark_w4a8_mxfp4_fp8.py (hunk 10 above)
Comment (divakar-amd) [reply vllm/model_executor/layers/quantization/quark/schemes/quark_w4a8_mxfp4_fp8.py]: done

File: vllm/model_executor/layers/quantization/quark/schemes/quark_w4a8_mxfp4_fp8.py (hunk 9 above)
Comment (divakar-amd) [reply vllm/model_executor/layers/quantization/quark/schemes/quark_w4a8_mxfp4_fp8.py]: done

File: vllm/model_executor/layers/quantization/quark/schemes/quark_w4a8_mxfp4_fp8.py (hunk 7 above)
Comment (divakar-amd) [reply vllm/model_executor/layers/quantization/quark/schemes/quark_w4a8_mxfp4_fp8.py:33]: Yes, via emulation. Tested on `gfx942` with a small linear-only model (llama-2-7b). Added new section "Testing emulation on gfx942" to PR summary for this.

File: vllm/model_executor/layers/quantization/quark/schemes/quark_w4a8_mxfp4_fp8.py (hunk 11 above)
Comment (divakar-amd) [reply vllm/model_executor/layers/quantization/quark/schemes/quark_w4a8_mxfp4_fp8.py:172]: Yes, via emulation. Tested on `gfx942` with a small linear-only model (llama-2-7b). Added new section "Testing emulation on gfx942" to PR summary for this.

**Review by tjtanaa — APPROVED** [review]
LGTM
